llvm/samples/clang/ClangMemoryMgmtExample.java

```diff
@@ -0,0 +1,229 @@
+		}
+	}
+
+	private static void withIndex(
+			final CXIndex index,
+			final Consumer<CXIndex> block
+	) {
+		try (final CXIndex ignored = index) {
+			try {
+				block.accept(index);
+			} finally {
+				clang_disposeIndex(index);
+			}
+		}
+	}
+
+	private static void withTranslationUnit(
+			final Supplier<CXTranslationUnit> lazyTranslationUnit,
+			final Consumer<CXTranslationUnit> block
+	) {
+		try (final CXTranslationUnit translationUnit = lazyTranslationUnit.get()) {
+			try {
+				block.accept(translationUnit);
+			} finally {
+				clang_disposeTranslationUnit(translationUnit);
+			}
+		}
+	}
+
+	private static void forEachToken(
+			final CXCursor cursor,
+			final Consumer<? super CXToken> action
+	) {
+		try (final CXSourceRange extent = clang_getCursorExtent(cursor)) {
+			try (final CXTranslationUnit translationUnit = clang_Cursor_getTranslationUnit(cursor)) {
+				try (final CXToken tokens = new CXToken()) {
+					final int[] tokenCountRef = new int[1];
+					clang_tokenize(translationUnit, extent, tokens, tokenCountRef);
+					final int tokenCount = tokenCountRef[0];
+					try {
+						IntStream.range(0, tokenCount)
+							 .mapToObj(tokens::position)
+							 .forEach(action);
+					} finally {
+						tokens.position(0L);
+						clang_disposeTokens(translationUnit, tokens, tokenCount);
+					}
+				}
+			}
+		}
+	}
+
+	private static final class AstVisitor extends CXCursorVisitor {
+		@Override
+		public int call(final CXCursor cursor, final CXCursor parent, final CXClientData clientData) {
+			try (final CXCursor c = cursor; final CXCursor p = parent; final CXClientData d = clientData) {
+				/*-
+				 * Entering a new `PointerScope` here is 100% necessary,
+				 * probably because the outer ("lower") stack frame is a
+				 * native one (i.e. `call()` is directly invoked by the
+				 * native code).
+				 *
+				 * Despite previously registered ("outer") pointer scopes
+				 * are still visible, having only a single scope per
+				 * translation unit (i.e., AST tree) rather than per cursor
+				 * eventually results in 100% usage of all CPU cores -- in
+				 * the native code.
+				 */
+				withPointerScope(() -> {
+					try (final CXString spelling = clang_getCursorKindSpelling(clang_getCursorKind(cursor))) {
+						out.println(spelling.getString());
+					}
+
+					try (final CXTranslationUnit translationUnit = clang_Cursor_getTranslationUnit(cursor)) {
+						forEachToken(cursor, token -> {
+							final TokenKind kind = TokenKind.valueOf(clang_getTokenKind(token));
+							try (final CXString spelling = clang_getTokenSpelling(translationUnit, token)) {
+								out.printf("\t%s(\"%s\")%n", kind, spelling.getString());
+							}
+						});
+					}
+
+				});
+			}
+
+			return CXChildVisit_Recurse;
+		}
+	}
+
+	private enum TokenKind {
+		Punctuation,
+
+		Keyword,
+
+		Identifier,
+
+		Literal,
+
+		Comment,
+		;
+
+		private static TokenKind valueOf(final int ordinal) {
+			return values()[ordinal];
+		}
+	}
+}
```

Comment:
Please convert tabs to spaces

Reply:
Done with https://github.com/bytedeco/javacpp-presets/pull/1522/commits/d903d6ba115f0ecd201f9da9e9024d5bfbc8da54.

I just followed the code style of the existing samples (which use tabs).

Reply:
They do? Haven't notice...